Add support for iOS builds

vnc0 opened the pull request:
This pull request introduces support for iOS builds by adding necessary flags and signing steps in the `GNUmakefile` and `GNUmakefile.llvm` files.

Key changes include:

### iOS Support:
* Added detection for iOS in the `GNUmakefile` and `GNUmakefile.llvm` files, setting appropriate flags when iOS is detected. [[1]](diffhunk://#diff-e3445fc75aa9c3e4a60fbe5394dcce12693022018216a8fbe0000fe9952850a6R73-R77) [[2]](diffhunk://#diff-df59777df6fde079ea6fa66de4cabe471ce9ddaff0d4e2e80fd0dd3558a10ff2R281-R282)
* Updated `frida_mode/GNUmakefile` to detect iOS and set appropriate architecture and flags. [[1]](diffhunk://#diff-183db3ca4941f790dec4c5bcf0d2a83834ec769f84121cc4c7071cf17482170eL22-R30) [[2]](diffhunk://#diff-183db3ca4941f790dec4c5bcf0d2a83834ec769f84121cc4c7071cf17482170eL98-R117)
* Added signing steps using `ldid` to ensure the binaries are properly signed for iOS. [[1]](diffhunk://#diff-e3445fc75aa9c3e4a60fbe5394dcce12693022018216a8fbe0000fe9952850a6R489-R575) [[2]](diffhunk://#diff-df59777df6fde079ea6fa66de4cabe471ce9ddaff0d4e2e80fd0dd3558a10ff2R444-R446) [[3]](diffhunk://#diff-183db3ca4941f790dec4c5bcf0d2a83834ec769f84121cc4c7071cf17482170eR404-R406) [[4]](diffhunk://#diff-183db3ca4941f790dec4c5bcf0d2a83834ec769f84121cc4c7071cf17482170eL425-R450)

### Entitlements:
* Added a new `entitlements.plist` file to specify the necessary entitlements for iOS binaries. My selection of entitlements is larger than currently needed, but this should not have a negative impact.
---
## Setup Instructions
### Environment
- **Tested On:** Jailbroken iPhone 12 mini (iOS 14.2.1, Taurine Jailbreak).
- **Alternative:** Cross-compile on macOS and transfer binaries to the device.
- **Recommendation:** For compiling on device a Procursus Jailbreak makes things a lot easier because we do not have to compile all the build tools and dependencies on our own.

### Install built tools / dependencies on target device
`sudo apt install wget git make cmake clang gawk llvm ldid coreutils build-essential xz-utils`

### Build Process
```
export CC=clang CXX=clang++
CFLAGS="-DUSEMMAP=1" IOS_SDK_PATH="/usr/share/SDKs/iPhoneOS.sdk" make
cd frida_mode
CFLAGS="-DUSEMMAP=1" make
```

## What works 

* Successful build and execution of default binaries (afl-fuzz, etc.). Solves https://github.com/AFLplusplus/AFLplusplus/issues/1869.
* Building with llvm-14 (LLVM_CONFIG=/usr/lib/llvm-14/bin/llvm-config) does succeed.
* Building Frida mode does succeed.

## What does not work (yet)

* Building with llvm-16 does fail.
* Both LLVM mode and Frida mode use DYLD_INSERT_LIBRARIES to inject the shared object files. This fails on iOS due to iOS sandbox restrictions (`could not load inserted library .. into hardened process because no suitable image found.  Did find: ... file system sandbox blocked mmap() of ...` ).

## Next Steps

- Investigate DYLD_INSERT_LIBRARIES sandbox bypass strategies. Help is appreciated.
- Add support for LLVM-16 on iOS.

vnc0 committed:
Add support for iOS builds

- Define IS_IOS variable and add conditional compilation flags for iOS
- Enable binary signing with ldid for iOS builds

vanhauser-thc commented:
looks good, thank you.
can you please:
1. fix the conflict
2. add how to compile for iOS to docs/INSTALL.md?
thanks!

vnc0 committed:
Add iOS cross-compilation support

- Updated CFLAGS and LDFLAGS.
- Disabled signing of bin2c during cross-compilation.

vnc0 committed:
Add iOS installation instructions

vnc0 committed:
Merge branch 'ios-cross' into ios

vnc0 committed:
Merge branch 'dev' into ios

vnc0 committed:
Update INSTALL.md

vanhauser-thc commented:
thank you!
can this be merged or are you still working on the next steps list?

vnc0 commented:
The current state can be merged. 

However, I will try to solve the problem with loading the shared object files on iOS in the next days. As I understand it at the moment, the problem is not to be solved with arbitrary permissions or by switching to dylibs, but is something more complicated. For example, Frida implements this [here](https://github.com/frida/frida-gum/blob/main/gum/backend-darwin/gumdarwinmapper.c). Other solutions are Cydia Substrate and Substitute. Because of the complexity I will probably just use Frida's solution here.
